File: src/main/java/com/timeback/ui/feature/permission/PermissionViewModel.java
package com.timeback.ui.feature.permission;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.timeback.ui.domain.gateway.FeatureGateway;
import com.timeback.ui.domain.model.ActionResult;
import com.timeback.ui.domain.model.ScreenState;

import javax.inject.Inject;
import dagger.hilt.android.lifecycle.HiltViewModel;

/**
 * UI-01 권한 화면 ViewModel.
 * 사용 권한 상태를 읽고 설정 화면을 열 수 있다.
 */
@HiltViewModel
public class PermissionViewModel extends ViewModel {

    private final FeatureGateway gateway;
    private final MutableLiveData<ScreenState> _state = new MutableLiveData<>(new ScreenState.Initial());
    public final LiveData<ScreenState> state = _state;

    private final MutableLiveData<ActionResult> _actionResult = new MutableLiveData<>();
    public final LiveData<ActionResult> actionResult = _actionResult;

    @Inject
    public PermissionViewModel(FeatureGateway gateway) {
        this.gateway = gateway;
        loadAccessState();
    }

    public void loadAccessState() {
        _state.setValue(new ScreenState.Loading());
        gateway.readAccessState().thenAccept(_state::postValue);
    }

    public void openSettings() {
        gateway.openUsageAccessSettings().thenAccept(_actionResult::postValue);
    }

    public void refresh() {
        gateway.refreshAccessState().thenAccept(_state::postValue);
    }
}

File: src/main/java/com/timeback/ui/feature/permission/PermissionViewModel.kt
package com.timeback.ui.feature.permission

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.timeback.ui.domain.gateway.FeatureGateway
import com.timeback.ui.domain.model.*
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * UI-01 권한·초기 진입 ViewModel
 * CT-04 §4: ReadAccessState, OpenUsageAccessSettings, RefreshAccessState
 */
@HiltViewModel
class PermissionViewModel @Inject constructor(
    private val gateway: FeatureGateway
) : ViewModel() {

    private val _state = MutableStateFlow<ScreenState>(ScreenState.Initial)
    val state: StateFlow<ScreenState> = _state.asStateFlow()

    init {
        loadAccessState()
    }

    fun loadAccessState() {
        viewModelScope.launch {
            _state.value = ScreenState.Loading
            _state.value = gateway.readAccessState()
        }
    }

    fun openSettings() {
        viewModelScope.launch {
            gateway.openUsageAccessSettings()
        }
    }

    fun refreshAccessState() {
        viewModelScope.launch {
            _state.value = ScreenState.Loading
            _state.value = gateway.refreshAccessState()
        }
    }
}
